core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -715,6 +715,20 @@ final class ZManaged[-R, +E, +A] private (reservation: ZIO[R, E, Reservation[R,
   def tap[R1 <: R, E1 >: E](f: A => ZManaged[R1, E1, Any]): ZManaged[R1, E1, A] =
     flatMap(a => f(a).as(a))
 
+  /**
+   * Returns an effect that effectfully peeks at the failure or success of the acquired resource.
+   */
+  final def tapBoth[R1 <: R, E1 >: E](f: E => ZManaged[R1, E1, Any])(g: A => ZManaged[R1, E1, Any])(
+    implicit ev: CanFail[E] // curried for better type inference
```

Comment:
This isn't curried on ZIO - would be great if we can keep things consistent.

Reply:
Also, I think you need to implement this using `foldM`, otherwise `g` will not be executed on error.

Reply:
Would this be an ok implementation then?
```scala
flatMap(a => foldM(f(_).as(a), a => g(a).as(a)))
```

Reply:
Unfortunately not - `flatMap` will only execute the provided function if the `this` instance succeeds.

You need something like:
```
foldM(
  e => f(e) *> ZManaged.fail(e),
  a => g(a).as(a)
)
```

Reply:
Just asking for my info, is `tap` written as: `flatMap(a => f(a).as(a))` OK because we're not peeking failures?

Reply:
Exactly right.